rio/transmat/impl/git2/git_warehouse.go

```diff
@@ -0,0 +1,130 @@
+	"go.polydawn.net/repeatr/api/def"
+	"go.polydawn.net/repeatr/rio"
+)
+
+/*
+	Refer to, interact with, and manage a warehouse.
+
+	Since this is git we're talking about, this is basically references
+	to another git repo.
+*/
+type Warehouse struct {
+	coord def.WarehouseCoord // user's string retained for messages
+	url   string
+}
+
+/*
+	Initialize a warehouse controller.
+
+	Note: parsing git URLs is *hard*.  This function is "best effort"
+	stuff and may not be able to error in all circumstances where the
+	git commands will error.  See `man gitremote-helpers` for one reason
+	this is the edge of a tarpit.  (This may improve as we do a better job
+	of pinning specific git versions and sandboxing their environment,
+	but at the moment, caveat emptor, and this is "PRs welcome" turf.)
+
+	May panic with:
+	  - Config Error: if the URI is unparsable or has an unsupported scheme.
+*/
+func NewWarehouse(coords rio.SiloURI) *Warehouse {
+	wh := &Warehouse{
+		coord: def.WarehouseCoord(coords),
+		url:   hammerRelativePaths(string(coords)),
+	}
+	return wh
+}
+
+/*
+	Desperately attempt to sanitize paths for git.
+
+	This is almost certainly flawed; but frankly, so is git's handling of
+	these things: it's not consistent across all commands (specifically,
+	`git clone` and `git ls-remote`).  We're going for best-effort here,
+	and slightly-better-than-not-trying -- low bars though those are.
+
+	Git has a MAJORLY hard time consistently understanding local paths; so,
+	we give up; internally all paths shall be absolutized because there's
+	simply no functional choice.
+
+	Here's a short list of issues we're concerned with:
+
+	  - There's no "--" in ls-remote, so in order to avoid ambiguity with
+	    arguments, we forbid things starting in "-".
+	  - `git ls-remote .` and `git ls-remote ./` are a special case and will
+	    return results even if the repo root is *above* your cwd, even
+	    though git clone will disagree for obvious reasons.
+	  - `git ls-remote ../sibling` works fine (and so does clone), but if
+	    you do it from *inside* the same repo, it just says no-such-repo
+	    (...but clone still works!)
+	  - `ls-remote` in particular is really willing to take you on a ride:
+	    if you have a repo-A, you make another git repo-B *inside* repo-A,
+	    then you can `ls-remote ..` correctly....
+	  - ....but if you then create a dir inside of repo-B, cd into it, and
+	    try `git ls-remote ../` again you'll get the info from **repo-A**,
+	    the grandparent.
+	  - So basically, relative paths in git simply Don't Work and are
+	    apparently Not Supported.
+
+	Beyond that, we mostly punt.  It's essentially impossible to validate
+	everything in advance in the same way that git will feel.
+	The area we're *really* minding here is where we've observed
+	`git clone` and `git ls-remote` behaviors to drift apart, because
+	that's exceptionally nasty to deal with or report about clearly.
+*/
+func hammerRelativePaths(coords string) string {
```

Comment:
We should be able to get rid of a looooot of this now that we're almost free of exec land...?